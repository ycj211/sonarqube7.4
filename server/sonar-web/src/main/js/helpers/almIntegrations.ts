/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { CurrentUser, AlmOrganization } from '../app/types';

export function hasAdvancedALMIntegration(user: CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string) {
  return almKey && almKey.startsWith('bitbucket');
}

export function isGithub(almKey?: string) {
  return almKey === 'github';
}

export function isVSTS(almKey?: string) {
  return almKey === 'microsoft';
}

export function isPersonal(organization?: AlmOrganization) {
  return Boolean(organization && organization.personal);
}

export function sanitizeAlmId(almKey?: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
