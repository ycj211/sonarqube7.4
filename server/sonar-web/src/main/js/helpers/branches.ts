/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';
import {
  BranchLike,
  Branch,
  BranchType,
  ShortLivingBranch,
  LongLivingBranch,
  PullRequest,
  MainBranch,
  BranchParameters
} from '../app/types';

export function isBranch(branchLike?: BranchLike): branchLike is Branch {
  return branchLike !== undefined && (branchLike as Branch).isMain !== undefined;
}

export function isShortLivingBranch(branchLike?: BranchLike): branchLike is ShortLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as ShortLivingBranch).type === BranchType.SHORT
  );
}

export function isLongLivingBranch(branchLike?: BranchLike): branchLike is LongLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as LongLivingBranch).type === BranchType.LONG
  );
}

export function isMainBranch(branchLike?: BranchLike): branchLike is MainBranch {
  return isBranch(branchLike) && branchLike.isMain;
}

export function isPullRequest(branchLike?: BranchLike): branchLike is PullRequest {
  return branchLike !== undefined && (branchLike as PullRequest).key !== undefined;
}

export function getPullRequestDisplayName(pullRequest: PullRequest) {
  return `${pullRequest.key} â€“ ${pullRequest.title}`;
}

export function getBranchLikeDisplayName(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? getPullRequestDisplayName(branchLike) : branchLike.name;
}

export function getBranchLikeKey(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? `pull-request-${branchLike.key}` : `branch-${branchLike.name}`;
}

export function getBranchQualityGateColor(status: string) {
  let indicatorColor = 'gray';
  if (status === 'ERROR') {
    indicatorColor = 'red';
  } else if (status === 'WARN') {
    indicatorColor = 'orange';
  } else if (status === 'OK') {
    indicatorColor = 'green';
  }
  return indicatorColor;
}

export function isSameBranchLike(a: BranchLike | undefined, b: BranchLike | undefined) {
  // main branches are always equal
  if (isMainBranch(a) && isMainBranch(b)) {
    return true;
  }

  // short- and long-living branches are compared by type and name
  if (
    (isLongLivingBranch(a) && isLongLivingBranch(b)) ||
    (isShortLivingBranch(a) && isShortLivingBranch(b))
  ) {
    return a.type === b.type && a.name === b.name;
  }

  // pull requests are compared by id
  if (isPullRequest(a) && isPullRequest(b)) {
    return a.key === b.key;
  }

  // finally if both parameters are `undefined`, consider them equal
  return a === b;
}

export function sortBranchesAsTree(branchLikes: BranchLike[]) {
  const result: BranchLike[] = [];

  const mainBranch = branchLikes.find(isMainBranch);
  const longLivingBranches = branchLikes.filter(isLongLivingBranch);
  const shortLivingBranches = branchLikes.filter(isShortLivingBranch);
  const pullRequests = branchLikes.filter(isPullRequest);

  // main branch is always first
  if (mainBranch) {
    result.push(
      mainBranch,
      ...getPullRequests(mainBranch.name),
      ...getNestedShortLivingBranches(mainBranch.name)
    );
  }

  // then all long-living branches
  sortBy(longLivingBranches, 'name').forEach(longLivingBranch => {
    result.push(
      longLivingBranch,
      ...getPullRequests(longLivingBranch.name),
      ...getNestedShortLivingBranches(longLivingBranch.name)
    );
  });

  // finally all orhpan pull requests and branches
  result.push(
    ...sortBy(pullRequests.filter(pr => pr.isOrphan), pullRequest => pullRequest.key),
    ...sortBy(shortLivingBranches.filter(branch => branch.isOrphan), branch => branch.name)
  );

  return result;

  /** Get all short-living branches (possibly nested) which should be merged to a given branch */
  function getNestedShortLivingBranches(mergeBranch: string) {
    const found: ShortLivingBranch[] = shortLivingBranches.filter(
      branch => branch.mergeBranch === mergeBranch
    );

    let i = 0;
    while (i < found.length) {
      const current = found[i];
      found.push(...shortLivingBranches.filter(branch => branch.mergeBranch === current.name));
      i++;
    }

    return sortBy(found, branch => branch.name);
  }

  function getPullRequests(base: string) {
    return sortBy(pullRequests.filter(pr => pr.base === base), pullRequest => pullRequest.key);
  }
}

export function getBranchLikeQuery(branchLike?: BranchLike): BranchParameters {
  if (isShortLivingBranch(branchLike) || isLongLivingBranch(branchLike)) {
    return { branch: branchLike.name };
  } else if (isPullRequest(branchLike)) {
    return { pullRequest: branchLike.key };
  } else {
    return {};
  }
}

// Create branch object from branch name or pull request key
export function fillBranchLike(
  branch?: string,
  pullRequest?: string
): ShortLivingBranch | PullRequest | undefined {
  if (branch) {
    return {
      isMain: false,
      mergeBranch: '',
      name: branch,
      type: BranchType.SHORT
    } as ShortLivingBranch;
  } else if (pullRequest) {
    return { base: '', branch: '', key: pullRequest, title: '' } as PullRequest;
  }
  return undefined;
}
