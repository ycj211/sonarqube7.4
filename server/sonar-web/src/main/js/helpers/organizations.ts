/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { Organization, OrganizationSubscription, CurrentUser } from '../app/types';

export function isPaidOrganization(organization: Organization | undefined): boolean {
  return Boolean(organization && organization.subscription === OrganizationSubscription.Paid);
}

export function hasPrivateAccess(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return (
    !isPaidOrganization(organization) ||
    isCurrentUserMemberOf(currentUser, organization, userOrganizations)
  );
}

export function isCurrentUserMemberOf(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return Boolean(
    organization &&
      isLoggedIn(currentUser) &&
      ((organization.actions && organization.actions.admin) ||
        userOrganizations.some(org => org.key === organization.key))
  );
}
