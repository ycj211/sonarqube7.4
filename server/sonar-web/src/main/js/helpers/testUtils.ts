/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl<C extends React.Component>(
  node: React.ReactElement<any>,
  options: ShallowRendererProps = {}
) {
  return shallow<C>(node, { ...options, context: { intl, ...options.context } });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}
