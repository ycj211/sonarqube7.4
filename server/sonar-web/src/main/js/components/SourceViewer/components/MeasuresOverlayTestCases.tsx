/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { orderBy } from 'lodash';
import MeasuresOverlayCoveredFiles from './MeasuresOverlayCoveredFiles';
import MeasuresOverlayTestCase from './MeasuresOverlayTestCase';
import { getTests } from '../../../api/tests';
import { TestCase, BranchLike } from '../../../app/types';
import { translate } from '../../../helpers/l10n';
import { getBranchLikeQuery } from '../../../helpers/branches';

interface Props {
  branchLike: BranchLike | undefined;
  componentKey: string;
}

interface State {
  loading: boolean;
  selectedTestId?: string;
  sort?: string;
  sortAsc?: boolean;
  testCases?: TestCase[];
}

export default class MeasuresOverlayTestCases extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = { loading: true };

  componentDidMount() {
    this.mounted = true;
    this.fetchTests();
  }

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.branchLike !== this.props.branchLike ||
      prevProps.componentKey !== this.props.componentKey
    ) {
      this.fetchTests();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchTests = () => {
    // TODO implement pagination one day...
    this.setState({ loading: true });
    getTests({
      ps: 500,
      testFileKey: this.props.componentKey,
      ...getBranchLikeQuery(this.props.branchLike)
    }).then(
      ({ tests: testCases }) => {
        if (this.mounted) {
          this.setState({ loading: false, testCases });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      }
    );
  };

  handleTestCasesSortClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    const { sort } = event.currentTarget.dataset;
    if (sort) {
      this.setState((state: State) => ({
        sort,
        sortAsc: sort === state.sort ? !state.sortAsc : true
      }));
    }
  };

  handleTestCaseClick = (selectedTestId: string) => {
    this.setState({ selectedTestId });
  };

  render() {
    const { selectedTestId, sort = 'name', sortAsc = true, testCases } = this.state;

    if (!testCases) {
      return null;
    }

    const selectedTest = testCases.find(test => test.id === selectedTestId);

    return (
      <div className="source-viewer-measures">
        <div className="source-viewer-measures-section source-viewer-measures-section-big">
          <div className="source-viewer-measures-card source-viewer-measures-card-fixed-height js-test-list">
            <div className="measures">
              <table className="source-viewer-tests-list">
                <tbody>
                  <tr>
                    <td className="source-viewer-test-status note" colSpan={3}>
                      {translate('component_viewer.measure_section.unit_tests')}
                      <br />
                      <span className="spacer-right">
                        {translate('component_viewer.tests.ordered_by')}
                      </span>
                      <a
                        className={classNames('js-sort-tests-by-duration', {
                          'active-link': sort === 'duration'
                        })}
                        data-sort="duration"
                        href="#"
                        onClick={this.handleTestCasesSortClick}>
                        {translate('component_viewer.tests.duration')}
                      </a>
                      <span className="slash-separator" />
                      <a
                        className={classNames('js-sort-tests-by-name', {
                          'active-link': sort === 'name'
                        })}
                        data-sort="name"
                        href="#"
                        onClick={this.handleTestCasesSortClick}>
                        {translate('component_viewer.tests.test_name')}
                      </a>
                      <span className="slash-separator" />
                      <a
                        className={classNames('js-sort-tests-by-status', {
                          'active-link': sort === 'status'
                        })}
                        data-sort="status"
                        href="#"
                        onClick={this.handleTestCasesSortClick}>
                        {translate('component_viewer.tests.status')}
                      </a>
                    </td>
                    <td className="source-viewer-test-covered-lines note">
                      {translate('component_viewer.covered_lines')}
                    </td>
                  </tr>
                  {sortTestCases(testCases, sort, sortAsc).map(testCase => (
                    <MeasuresOverlayTestCase
                      key={testCase.id}
                      onClick={this.handleTestCaseClick}
                      testCase={testCase}
                    />
                  ))}
                </tbody>
              </table>
            </div>
          </div>
        </div>
        {selectedTest && <MeasuresOverlayCoveredFiles testCase={selectedTest} />}
      </div>
    );
  }
}

function sortTestCases(testCases: TestCase[], sort: string, sortAsc: boolean) {
  const mainOrder = sortAsc ? 'asc' : 'desc';
  if (sort === 'duration') {
    return orderBy(testCases, ['durationInMs', 'name'], [mainOrder, 'asc']);
  } else if (sort === 'status') {
    return orderBy(testCases, ['status', 'name'], [mainOrder, 'asc']);
  } else {
    return orderBy(testCases, ['name'], [mainOrder]);
  }
}
