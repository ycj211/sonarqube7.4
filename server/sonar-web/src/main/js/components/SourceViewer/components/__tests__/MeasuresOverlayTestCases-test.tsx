/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import MeasuresOverlayTestCases from '../MeasuresOverlayTestCases';
import { waitAndUpdate, click } from '../../../../helpers/testUtils';
import { ShortLivingBranch, BranchType } from '../../../../app/types';

jest.mock('../../../../api/tests', () => ({
  getTests: () =>
    Promise.resolve({
      tests: [
        {
          id: 'AWGub2mGGZxsAttCZwQy',
          name: 'testAdd_WhichFails',
          fileKey: 'test:fake-project-for-tests:src/test/java/bar/SimplestTest.java',
          fileName: 'src/test/java/bar/SimplestTest.java',
          status: 'FAILURE',
          durationInMs: 6,
          coveredLines: 3,
          message: 'expected:<9> but was:<2>',
          stacktrace:
            'java.lang.AssertionError: expected:<9> but was:<2>\n\tat org.junit.Assert.fail(Assert.java:93)\n\tat org.junit.Assert.failNotEquals(Assert.java:647)'
        },
        {
          id: 'AWGub2mGGZxsAttCZwQz',
          name: 'testAdd_InError',
          fileKey: 'test:fake-project-for-tests:src/test/java/bar/SimplestTest.java',
          fileName: 'src/test/java/bar/SimplestTest.java',
          status: 'ERROR',
          durationInMs: 2,
          coveredLines: 3
        },
        {
          id: 'AWGub2mFGZxsAttCZwQx',
          name: 'testAdd',
          fileKey: 'test:fake-project-for-tests:src/test/java/bar/SimplestTest.java',
          fileName: 'src/test/java/bar/SimplestTest.java',
          status: 'OK',
          durationInMs: 8,
          coveredLines: 3
        }
      ]
    })
}));

const branchLike: ShortLivingBranch = {
  isMain: false,
  mergeBranch: 'master',
  name: 'feature',
  type: BranchType.SHORT
};

it('should render', async () => {
  const wrapper = shallow(
    <MeasuresOverlayTestCases branchLike={branchLike} componentKey="component-key" />
  );
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('.js-sort-tests-by-duration'), {
    currentTarget: { blur() {}, dataset: { sort: 'duration' } }
  });
  expect(wrapper).toMatchSnapshot();
});
