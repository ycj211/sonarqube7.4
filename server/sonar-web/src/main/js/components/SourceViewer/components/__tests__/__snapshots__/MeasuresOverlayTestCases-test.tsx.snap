// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="source-viewer-measures"
>
  <div
    className="source-viewer-measures-section source-viewer-measures-section-big"
  >
    <div
      className="source-viewer-measures-card source-viewer-measures-card-fixed-height js-test-list"
    >
      <div
        className="measures"
      >
        <table
          className="source-viewer-tests-list"
        >
          <tbody>
            <tr>
              <td
                className="source-viewer-test-status note"
                colSpan={3}
              >
                component_viewer.measure_section.unit_tests
                <br />
                <span
                  className="spacer-right"
                >
                  component_viewer.tests.ordered_by
                </span>
                <a
                  className="js-sort-tests-by-duration"
                  data-sort="duration"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.duration
                </a>
                <span
                  className="slash-separator"
                />
                <a
                  className="js-sort-tests-by-name active-link"
                  data-sort="name"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.test_name
                </a>
                <span
                  className="slash-separator"
                />
                <a
                  className="js-sort-tests-by-status"
                  data-sort="status"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.status
                </a>
              </td>
              <td
                className="source-viewer-test-covered-lines note"
              >
                component_viewer.covered_lines
              </td>
            </tr>
            <MeasuresOverlayTestCase
              key="AWGub2mFGZxsAttCZwQx"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 8,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mFGZxsAttCZwQx",
                  "name": "testAdd",
                  "status": "OK",
                }
              }
            />
            <MeasuresOverlayTestCase
              key="AWGub2mGGZxsAttCZwQz"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 2,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mGGZxsAttCZwQz",
                  "name": "testAdd_InError",
                  "status": "ERROR",
                }
              }
            />
            <MeasuresOverlayTestCase
              key="AWGub2mGGZxsAttCZwQy"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 6,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mGGZxsAttCZwQy",
                  "message": "expected:<9> but was:<2>",
                  "name": "testAdd_WhichFails",
                  "stacktrace": "java.lang.AssertionError: expected:<9> but was:<2>
	at org.junit.Assert.fail(Assert.java:93)
	at org.junit.Assert.failNotEquals(Assert.java:647)",
                  "status": "FAILURE",
                }
              }
            />
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render 2`] = `
<div
  className="source-viewer-measures"
>
  <div
    className="source-viewer-measures-section source-viewer-measures-section-big"
  >
    <div
      className="source-viewer-measures-card source-viewer-measures-card-fixed-height js-test-list"
    >
      <div
        className="measures"
      >
        <table
          className="source-viewer-tests-list"
        >
          <tbody>
            <tr>
              <td
                className="source-viewer-test-status note"
                colSpan={3}
              >
                component_viewer.measure_section.unit_tests
                <br />
                <span
                  className="spacer-right"
                >
                  component_viewer.tests.ordered_by
                </span>
                <a
                  className="js-sort-tests-by-duration active-link"
                  data-sort="duration"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.duration
                </a>
                <span
                  className="slash-separator"
                />
                <a
                  className="js-sort-tests-by-name"
                  data-sort="name"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.test_name
                </a>
                <span
                  className="slash-separator"
                />
                <a
                  className="js-sort-tests-by-status"
                  data-sort="status"
                  href="#"
                  onClick={[Function]}
                >
                  component_viewer.tests.status
                </a>
              </td>
              <td
                className="source-viewer-test-covered-lines note"
              >
                component_viewer.covered_lines
              </td>
            </tr>
            <MeasuresOverlayTestCase
              key="AWGub2mGGZxsAttCZwQz"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 2,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mGGZxsAttCZwQz",
                  "name": "testAdd_InError",
                  "status": "ERROR",
                }
              }
            />
            <MeasuresOverlayTestCase
              key="AWGub2mGGZxsAttCZwQy"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 6,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mGGZxsAttCZwQy",
                  "message": "expected:<9> but was:<2>",
                  "name": "testAdd_WhichFails",
                  "stacktrace": "java.lang.AssertionError: expected:<9> but was:<2>
	at org.junit.Assert.fail(Assert.java:93)
	at org.junit.Assert.failNotEquals(Assert.java:647)",
                  "status": "FAILURE",
                }
              }
            />
            <MeasuresOverlayTestCase
              key="AWGub2mFGZxsAttCZwQx"
              onClick={[Function]}
              testCase={
                Object {
                  "coveredLines": 3,
                  "durationInMs": 8,
                  "fileKey": "test:fake-project-for-tests:src/test/java/bar/SimplestTest.java",
                  "fileName": "src/test/java/bar/SimplestTest.java",
                  "id": "AWGub2mFGZxsAttCZwQx",
                  "name": "testAdd",
                  "status": "OK",
                }
              }
            />
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
